src/crypto/openssl/cose_sign.cpp

```diff
@@ -0,0 +1,142 @@
+    {
+      QCBOREncode_AddSZStringToMapN(encode_ctx, label, value.c_str());
+    }
+
+    QCBOREncode_CloseMap(encode_ctx);
+    QCBOREncode_CloseBstrWrap2(encode_ctx, false, &ctx->protected_parameters);
+  }
+
+  /* The original `t_cose_sign1_encode_parameters` can't accept a custom set of
+   parameters to be encoded into headers. This version tags the context as
+   COSE_SIGN1 and encodes the protected headers in the following order:
+     - defaults
+       - algorithm version
+     - those provided by caller
+   */
+  void encode_parameters_custom(
+    struct t_cose_sign1_sign_ctx* me,
+    QCBOREncodeContext* cbor_encode,
+    const std::unordered_map<int64_t, std::string>& protected_headers)
+  {
+    QCBOREncode_AddTag(cbor_encode, CBOR_TAG_COSE_SIGN1);
+    QCBOREncode_OpenArray(cbor_encode);
+
+    encode_protected_headers(me, cbor_encode, protected_headers);
+
+    QCBOREncode_OpenMap(cbor_encode);
+    // Explicitly leave unprotected headers empty to be an empty map.
+    QCBOREncode_CloseMap(cbor_encode);
+  }
+}
+
+namespace ccf::crypto
+{
+  std::vector<uint8_t> cose_sign1(
+    EVP_PKEY* key,
+    const std::unordered_map<int64_t, std::string>& protected_headers,
+    std::span<const uint8_t> payload)
+  {
+    const auto buf_size = estimate_buffer_size(protected_headers, payload);
+    Q_USEFUL_BUF_MAKE_STACK_UB(signed_cose_buffer, buf_size);
+
+    QCBOREncodeContext cbor_encode;
+    QCBOREncode_Init(&cbor_encode, signed_cose_buffer);
+
+    t_cose_sign1_sign_ctx sign_ctx;
+    t_cose_sign1_sign_init(&sign_ctx, 0, T_COSE_ALGORITHM_ES256);
+
+    t_cose_key signing_key;
+    signing_key.crypto_lib = T_COSE_CRYPTO_LIB_OPENSSL;
+    signing_key.k.key_ptr = key;
+
+    t_cose_sign1_set_signing_key(&sign_ctx, signing_key, NULL_Q_USEFUL_BUF_C);
+
+    encode_parameters_custom(&sign_ctx, &cbor_encode, protected_headers);
+
+    // Mark empty payload manually.
+    QCBOREncode_AddNULL(&cbor_encode);
+
+    // If payload is empty - we still want to sign. Putting NULL_Q_USEFUL_BUF_C,
+    // however, makes t_cose think that the payload is included into the
+    // context. Luckily, passing empty string instead works, so t_cose works
+    // emplaces it for TBS (to be signed) as an empty byte sequence.
+    q_useful_buf_c payload_to_encode = {"", 0};
+    if (!payload.empty())
+    {
+      payload_to_encode.ptr = payload.data();
+      payload_to_encode.len = payload.size();
+    }
+    auto err = t_cose_sign1_encode_signature_aad_internal(
+      &sign_ctx, NULL_Q_USEFUL_BUF_C, payload_to_encode, &cbor_encode);
+    if (err)
+    {
+      throw COSESignError(
+        fmt::format("Can't encode signature with error code {}", err));
+    }
+
+    struct q_useful_buf_c signed_cose;
+    auto qerr = QCBOREncode_Finish(&cbor_encode, &signed_cose);
+    if (qerr)
+    {
+      throw COSESignError(
+        fmt::format("Can't finish QCBOR encoding with error code {}", err));
+    }
+
+    std::vector<uint8_t> cose_sign(signed_cose.len);
```

Comment:
Can we do something like:

```suggestion
    std::vector<uint8_t> cose_sign(signed_cose.ptr, signed_cose.ptr + signed_cose.len);
```